From 7c14788ed8a9b64d45933adc71fca2a0007ba01a Mon Sep 17 00:00:00 2001
From: Myron Stowe <myron.stowe@redhat.com>
Date: Sat, 19 Jul 2014 04:25:33 -0400
Subject: [pci] Add pci_bus_address() to get bus address of a BAR

Message-id: <20140719042533.19438.32481.stgit@gir.stowe>
Patchwork-id: 85989
O-Subject: [RHEL7.1 PATCH 27/81] PCI: Add pci_bus_address() to get bus address of a BAR
Bugzilla: 1110895
RH-Acked-by: Tony Camuso <tcamuso@redhat.com>
RH-Acked-by: Prarit Bhargava <prarit@redhat.com>
RH-Acked-by: Stefan Assmann <sassmann@redhat.com>

Bugzilla: https://bugzilla.redhat.com/show_bug.cgi?id=1110895
Upstream Status: 06cf56e497c8c1469b0931caa7d5b1d827655fe2

commit 06cf56e497c8c1469b0931caa7d5b1d827655fe2
Author: Bjorn Helgaas <bhelgaas@google.com>
Date:   Sat Dec 21 08:33:26 2013 -0700

    PCI: Add pci_bus_address() to get bus address of a BAR

    We store BAR information as a struct resource, which contains the CPU
    address, not the bus address.  Drivers often need the bus address, and
    there's currently no convenient way to get it, so they often read the
    BAR directly, or use the resource address (which doesn't work if there's
    any translation between CPU and bus addresses).

    Add pci_bus_address() to make this convenient.

    Signed-off-by: Bjorn Helgaas <bhelgaas@google.com>
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/include/linux/pci.h b/include/linux/pci.h
index 6f15d61..0853e58 100644
--- a/include/linux/pci.h
+++ b/include/linux/pci.h
@@ -1109,6 +1109,14 @@ int __must_check pci_bus_alloc_resource(struct pci_bus *bus,
         resource_size_t),
    void *alignf_data);
 
+static inline dma_addr_t pci_bus_address(struct pci_dev *pdev, int bar)
+{
+ struct pci_bus_region region;
+
+ pcibios_resource_to_bus(pdev->bus, &region, &pdev->resource[bar]);
+ return region.start;
+}
+
 /* Proper probing supporting hot-pluggable devices */
 int __must_check __pci_register_driver(struct pci_driver *, struct module *,
            const char *mod_name);
-- 
1.7.1